### PR TITLE
Datadog improvements

### PR DESCRIPTION
**Pull request by Will-Howard**
In this PR:
 - enable Real User Monitoring (RUM), which enables us to essentially see a log of the console and network tab from all sessions
 - enable Session Replay, which lets us see a video-style playback of user session. Unfortunately the css isn't being captured correctly at the moment so the video comes out all messed up. I think this might be a result of it trying to access the stylesheet on localhost so I would like to deploy this to staging to test anyway. If this doesn't work the [troubleshooting docs](https://docs.datadoghq.com/real_user_monitoring/session_replay/troubleshooting/) are not very helpful so we may have to go without this for now :(
 - add some user info to backend traces

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203549116634243) by [Unito](https://www.unito.io)

### CODE REVIEW

File: build.js [hunk 1]
```diff
@@ -91,6 +91,7 @@ const bundleDefinitions = {
   "defaultSiteAbsoluteUrl": `\"${process.env.ROOT_URL || ""}\"`,
   "buildId": `"${latestCompletedBuildId}"`,
   "serverPort": getServerPort(),
+  "ddEnv": `\"${process.env.DD_ENV || "local"}\"`,
```
Comment (Will-Howard) [inline build.js:94]: This will be "staging" or "prod"

File: packages/lesswrong/client/datadogRum.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  } : {});
+
+  if (user && !user.allowDatadogSessionReplay) {
+    // eslint-disable-next-line no-console
+    console.log("Session Replay disabled")
+    datadogRum.stopSessionReplayRecording();
+  } else {
+    // eslint-disable-next-line no-console
+    console.log("Session Replay enabled")
+    datadogRum.startSessionReplayRecording();
```
Comment (Will-Howard) [inline packages/lesswrong/client/datadogRum.ts:41]: This is idempotent so it doesn't matter if it gets called multiple times

File: packages/lesswrong/client/logging.ts [hunk 3]
```diff
@@ -39,23 +40,25 @@ if (sentryUrl && sentryEnvironment && sentryRelease) {
 
 // Initializing sentry on the client browser
 
-userIdentifiedCallback.add(function identifyUserToSentry(user: UsersCurrent) {
-  // Set user in sentry scope
+userChangedCallback.add(function identifyUserToSentry(user: UsersCurrent | null) {
+  // Set user in sentry scope, or clear user if they have logged out
   Sentry.configureScope((scope) => {
-    scope.setUser({id: user._id, email: getUserEmail(user), username: user.username});
+    scope.setUser(user ? {id: user._id, email: getUserEmail(user), username: user.username} : null);
```
Comment (Will-Howard) [inline packages/lesswrong/client/logging.ts:46]: There was a bug in these methods previously. `userIdentifiedCallback` would only be called if there was a currentUser so if a user logged out they wouldn't be unset in sentry or google analytics

File: packages/lesswrong/server/datadog/datadogMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ * - Allow the headers required for Real User Monitoring
+ */
+export const datadogMiddleware = (req, res, next) => {
+  const span = tracer.scope().active()
+  if (span !== null) {
+    // @ts-ignore - there is currently no public API for getting the root span, this is the accepted way (see https://github.com/DataDog/dd-trace-js/issues/725#issuecomment-805277510)
+    const rootSpan = span.context()._trace.started[0]
+    const user = req.user
+    if (user) {
+      rootSpan.setTag('usr', {id: user._id, name: user.slug, email: user.email})
```
Comment (Will-Howard) [inline packages/lesswrong/server/datadog/datadogMiddleware.ts]: Add the user to the backend traces

File: packages/lesswrong/server/datadog/datadogMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * - Allow the headers required for Real User Monitoring
+ */
+export const datadogMiddleware = (req, res, next) => {
+  const span = tracer.scope().active()
+  if (span !== null) {
+    // @ts-ignore - there is currently no public API for getting the root span, this is the accepted way (see https://github.com/DataDog/dd-trace-js/issues/725#issuecomment-805277510)
+    const rootSpan = span.context()._trace.started[0]
+    const user = req.user
+    if (user) {
+      rootSpan.setTag('usr', {id: user._id, name: user.displayName, email: user.email, slug: user.slug})
```
Comment (Will-Howard) [inline packages/lesswrong/server/datadog/datadogMiddleware.ts:14]: `usr: {id, name, email}` is the preferred format for user info in datadog 🤷‍♂️

File: packages/lesswrong/server/manualMigrations/2022-12-13-defaultSessionReplay.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  name: "defaultSessionReplay",
+  dateWritten: "2022-12-13",
+  idempotent: true,
+  action: async () => {
+    await fillDefaultValues({
+      collection: Users,
+      fieldName: "allowDatadogSessionReplay",
+    });
+  },
+});
```
Comment (s-cheng) [inline packages/lesswrong/server/manualMigrations/2022-12-13-defaultSessionReplay.ts:15]: Do we need this if the field is optional and the default value is `false` anyway?

File: packages/lesswrong/server/datadog/datadogMiddleware.ts (hunk 5 above)
Comment (s-cheng) [reply packages/lesswrong/server/datadog/datadogMiddleware.ts:14]: Does this need to match the user data in `configureDatadogRum()`? The `email` values could be different. Also, if they do need to be kept in sync, could you add comments in the code explaining that?

**Review by s-cheng — APPROVED** [review]
Cool, seems good! Do you have a plan around how to get people to opt-in to session replay? (since it will be defaulted to `false` for everyone)

File: packages/lesswrong/server/manualMigrations/2022-12-13-defaultSessionReplay.ts (hunk 6 above)
Comment (Will-Howard) [reply packages/lesswrong/server/manualMigrations/2022-12-13-defaultSessionReplay.ts:15]: I guess not, I'll remove it

**Comment by Will-Howard** [comment]
> Cool, seems good! Do you have a plan around how to get people to opt-in to session replay? (since it will be defaulted to `false` for everyone)

I'm going to put it in slack, and then maybe we can try to encourage people to turn it on in a forum updates post. I wouldn't mind if not that many people turn it on though as it does feel a bit intrusive

File: packages/lesswrong/client/datadogRum.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    "https://forum-staging.effectivealtruism.org"
+    // TODO add LW domains here if they want to use datadog
+  ]
+});
+
+export function configureDatadogRum(user: UsersCurrent | UsersEdit | DbUser | null) {
+  if (forumTypeSetting.get() !== 'EAForum') return
+
+  // Set the user which will appear in traces. This info should match what
+  datadogRum.setUser(user ? getDatadogUser(user) : {});
```
Comment (s-cheng) [inline packages/lesswrong/client/datadogRum.ts:30]: @Will-Howard FYI I think you forgot to finish this comment

**Pull request merged**